kevlar/call.py

```diff
@@ -13,57 +13,118 @@
-    if len(diffs) == 1:
-        localcoord = offset + diffs[0][0]
-        refr = diffs[0][1].upper()
-        alt = diffs[0][2].upper()
-        globalregex = re.search('(\S+)_(\d+)-(\d+)', target.name)
-        assert globalregex, target.name
-        seqid = globalregex.group(1)
-        globaloffset = int(globalregex.group(2))
-        globalcoord = globaloffset + localcoord
-        return '{:s}:{:d}:{:s}->{:s}'.format(seqid, globalcoord, refr, alt)
+    if len(diffs) == 0:
+        raise StopIteration
+
+    snvs = list()
+    for diff in diffs:
+        minpos = max(diff[0] - ksize + 1, 0)
+        maxpos = min(diff[0] + ksize, length)
+        window = q[minpos:maxpos]
+        numoverlappingkmers = len(window) - ksize + 1
+        kmers = [window[i:i+ksize] for i in range(numoverlappingkmers)]
+
+        refr = diff[1].upper()
+        alt = diff[2].upper()
+        localcoord = offset + diff[0]
+        seqid, globalcoord = local_to_global(localcoord, target.name)
+        snv = VariantSNV(seqid, globalcoord, refr, alt, window=window)
+        snvs.append(snv)
+    return snvs
 
 
 def call_deletion(target, query, offset, leftmatch, indellength):
+    refr = target.sequence[offset+leftmatch-1:offset+leftmatch+indellength]
+    alt = refr[0]
+    assert len(refr) == indellength + 1
     localcoord = offset + leftmatch
-    globalregex = re.search('(\S+)_(\d+)-(\d+)', target.name)
-    assert globalregex, target.name
-    seqid = globalregex.group(1)
-    globaloffset = int(globalregex.group(2))
-    globalcoord = globaloffset + localcoord
-    return '{:s}:{:d}:{:d}D'.format(seqid, globalcoord, indellength)
+    seqid, globalcoord = local_to_global(localcoord, target.name)
+    return [VariantIndel(seqid, globalcoord - 1, refr, alt)]
 
 
 def call_insertion(target, query, offset, leftmatch, indellength):
-    insertion = query.sequence[leftmatch:leftmatch+indellength]
-    assert len(insertion) == indellength
+    insertion = query.sequence[leftmatch-1:leftmatch+indellength]
+    refr = insertion[0]
+    assert len(insertion) == indellength + 1
     localcoord = offset + leftmatch
-    globalregex = re.search('(\S+)_(\d+)-(\d+)', target.name)
-    assert globalregex, target.name
-    seqid = globalregex.group(1)
-    globaloffset = int(globalregex.group(2))
-    globalcoord = globaloffset + localcoord
-    return '{:s}:{:d}:I->{:s}'.format(seqid, globalcoord, insertion)
+    seqid, globalcoord = local_to_global(localcoord, target.name)
+    return [VariantIndel(seqid, globalcoord - 1, refr, insertion)]
 
 
-def make_call(target, query, cigar):
+def make_call(target, query, cigar, ksize):
     snvmatch = re.search('^(\d+)D(\d+)M(\d+)D$', cigar)
     snvmatch2 = re.search('^(\d+)D(\d+)M(\d+)D(\d+)M$', cigar)
     if snvmatch:
         offset = int(snvmatch.group(1))
         length = int(snvmatch.group(2))
-        return call_snv(target, query, offset, length)
+        return call_snv(target, query, offset, length, ksize)
```

Comment:
Looking at the tests, I think it's pretty clear this function's return value(s) need rethinking. Perhaps I should have the `Variant` base class handle "no call" scenarios and have these call functions all return a single object of class `Variant`.